### PR TITLE
[Feature] Megatron LayerNorm sequence parallelism (--enable-layernorm-sp)

### PR DESCRIPTION
**Pull request by MartinHua**
# [Feature] Megatron LayerNorm sequence parallelism (`--enable-layernorm-sp`)

## What

Megatron-style LayerNorm sequence parallelism ([arXiv:2205.05198](https://arxiv.org/abs/2205.05198)) for prefill under pure tensor parallelism: the row-parallel `all_reduce` becomes `reduce_scatter` (g-bar) + `all_gather` (g), so LayerNorm/residual run on `1/tp` of the tokens -- less activation memory, and no extra communication volume, since the two forms move the same bytes. Both collectives are fused with the adjacent GEMMs via torch symmetric memory. Off by default and byte-identical when off.

## Design

All SP logic lives in the new `layers/layernorm_sp.py`. There are **no changes under `srt/models/`** and none to `dp_attention.py`.

- `LayerCommunicator` owns the region: it entry-scatters on the first decoder layer, then delegates `prepare_attn` / `prepare_mlp` / `postprocess_layer` to an all-`SCATTERED` sibling.
- `Row`/`ColumnParallelLinear` auto-participate in g / g-bar while the region is active, so no per-model marking.
- `LogitsProcessor` exit-gathers before the LM head.
- `flags.sp.enabled` is materialized once per worker by `initialize_layernorm_sp`, beside `initialize_dp_attention`.
- Enabling another family is one line in `SP_SUPPORTED_ARCHITECTURES` plus validation on hardware.

The exit gather recomputes its condition from config + `forward_mode` rather than reading the per-forward `sp_active` flag: that flag is written by Python *inside* the CUDA-graph-captured region, so it is stale on replay, and reading it there fed sequence-sharded hidden states to the LM head.

## Requirements

`arg_groups/layernorm_sp_hook.py` rejects these at startup, from the resolution pipeline: an architecture outside `SP_SUPPORTED_ARCHITECTURES` (`Qwen3ForCausalLM`; others lack the `LayerCommunicator` and participant-linear wiring), `tp_size <= 1`, `--enable-dp-attention` (SP shards over the full TP group, which under DP attention spans groups holding different sequences), and speculative decoding (the captured aux hidden states would be sequence-sharded).

Silent gates, not errors: prefill (`EXTEND`) only, and the fused path needs unquantized bias-free bf16/fp16 linears, falling back to plain collectives otherwise. NVLink/NVSwitch is required in practice but not yet enforced.

## Tests

`test/registered/unit/layers/test_layernorm_sp.py` (CPU) covers the allowlist, the prefill-only rule, the four guards, and a regression case pinning that the exit-gather predicate ignores `sp_active`. `test/registered/models_e2e/test_layernorm_sp.py` (2 GPU) runs Qwen3-8B with `--tp 2 --enable-layernorm-sp` against a GSM8K threshold.

## Results

GSM8K is Qwen3-8B, tp=2, 1319 questions; prefill latency is the same model at 16K. Accuracy deltas are fp-reordering noise.

- **2xH100** (this revision): GSM8K 0.904 off / 0.906 on; prefill 0.28173 -> **0.27180 s (-3.5%)**.
- **2xB200** (sm100, torch 2.13): GSM8K 0.909 off / 0.908 on; prefill 0.14388 -> **0.13589 s (-5.6%)**. The fused symmetric-memory kernels are numerically correct on sm100 and are what delivers the win there -- forced off, SP measures 0.1421-0.1442 s, neutral against baseline. An FP8 checkpoint (quantized, so the plain-collective fallback) also serves correctly.
- **8xH100, an earlier revision of this branch**: **-16.5% prefill latency and -27.7% activation memory** (Qwen3-32B, 32K), and no recompile storm under piecewise-prefill compile (44 vs 40 baseline). Not re-run on the current revision.

The gain scales with TP degree.

## Follow-ups

- More model families (one allowlist line each, plus validation).
- B300 (B200 / sm100 is covered above; both report capability major 10, so the same code paths apply).
- Further out: DP-attention-aware SP, speculative decoding, an explicit NVLink precondition, chunked/split-prefill.

---
### CI States

Latest PR Test (Base): :x: [Run #33325891316](https://github.com/sgl-project/sglang/actions/runs/33325891316)
Latest PR Test (Extra): :x: [Run #33325891165](https://github.com/sgl-project/sglang/actions/runs/33325891165)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33325891318](https://github.com/sgl-project/sglang/actions/runs/33325891318)

### CODE REVIEW

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -962,6 +962,15 @@ class ServerArgs:
         bool,
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py:1113]: Please add a unit test and an e2e test for layernorm sp

File: python/sglang/benchmark/one_batch.py [hunk 2]
```diff
@@ -742,10 +742,17 @@ def latency_test_run_once(
         )
```
Comment (Fridge003) [inline python/sglang/benchmark/one_batch.py:791]: Don't change this file

File: python/sglang/srt/layers/dp_attention.py [hunk 3]
```diff
@@ -250,6 +250,72 @@ def get_is_extend_in_batch() -> bool:
     return get_forward().is_extend_in_batch
 
 
+# --- Megatron LayerNorm sequence parallelism (arXiv:2205.05198) ---
```
Comment (Fridge003) [inline python/sglang/srt/layers/dp_attention.py]: Can we open a single file for layernorm_sp, and put all the related classes/util functions there. 
For example, `python/sglang/srt/layers/layernorm_sp.py`

The `dp_attention.py` file shouldn't be modified

File: python/sglang/srt/models/qwen2.py [hunk 4]
```diff
@@ -27,7 +27,12 @@
     get_pp_indices,
```
Comment (Fridge003) [inline python/sglang/srt/models/qwen2.py:27]: Can we fully decouple the implementation of modeling logics and sp?
Ideally there should be no change in files under python/sglang/srt/models

**Comment by Fridge003** [comment]
Please solve the conflicts

File: python/sglang/benchmark/one_batch.py (hunk 2 above)
Comment (MartinHua) [reply python/sglang/benchmark/one_batch.py:791]: Reverted

File: python/sglang/srt/layers/dp_attention.py (hunk 3 above)
Comment (MartinHua) [reply python/sglang/srt/layers/dp_attention.py]: Done. Everything SP now lives in `python/sglang/srt/layers/layernorm_sp.py`: the supported-architecture allowlist, config validation, the per-forward `sp_active` flag, the entry-scatter / exit-gather collectives, and the fused g / ḡ matmul helpers (torch symmetric memory, with plain-collective fallbacks).

`dp_attention.py` is fully reverted — zero changes in this PR.

File: python/sglang/srt/models/qwen2.py (hunk 4 above)
Comment (MartinHua) [reply python/sglang/srt/models/qwen2.py:27]: Done — there are now **no changes under `python/sglang/srt/models/`**. The whole diff is 8 files: the new `layers/layernorm_sp.py`, plus `communicator.py`, `linear.py`, `logits_processor.py`, `runtime_context.py`, `server_args.py`, and the two tests.

How the decoupling works:

- `LayerCommunicator` owns the SP region: on the first decoder layer (`residual is None`) it re-evaluates whether this forward runs SP and does the entry scatter, then delegates `prepare_attn` / `prepare_mlp` / `postprocess_layer` to an all-`SCATTERED` sibling while the region is active.
- `Row`/`ColumnParallelLinear` auto-participate in g / ḡ while the region is active, so no per-model `sequence_parallel` marking is needed.
- The exit gather lives in `LogitsProcessor`, which is model-agnostic.
- Which models are enabled is a one-line allowlist in `layernorm_sp.py` (`Qwen3ForCausalLM` today); adding a family is that line plus hardware validation.

Validation sits in `ServerArgs.check_server_args` rather than in the layers, so a model that never builds a `LayerCommunicator` (e.g. Qwen2) fails loud instead of silently ignoring the flag.

File: python/sglang/srt/server_args.py (hunk 1 above)
Comment (MartinHua) [reply python/sglang/srt/server_args.py:1113]: Added both.

- `test/registered/unit/layers/test_layernorm_sp.py` (no server): allowlist gating, the prefill-only activation rule, the four config guards (supported architecture / `tp_size > 1` / not `--enable-dp-attention` / not speculative decoding), and a regression case pinning that the exit-gather predicate stays independent of the `sp_active` flag.
- `test/registered/models_e2e/test_layernorm_sp.py`: Qwen3-1.7B with `--tp 2 --enable-layernorm-sp`, asserting a GSM8K threshold.

Both pass on 2×H100: unit 9/9, e2e in ~67 s. Registered with the `stage=` / `runner_config=` form (`base-b` / `1-gpu-small` and `2-gpu-large`).

File: test/registered/unit/layers/test_layernorm_sp.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+from types import SimpleNamespace
+from unittest.mock import patch
+
+from sglang.srt.layers import layernorm_sp
+from sglang.srt.model_executor.forward_batch_info import ForwardMode
+from sglang.srt.runtime_context import reset_context
+from sglang.test.ci.ci_register import register_cuda_ci
+from sglang.test.test_utils import CustomTestCase
+
+register_cuda_ci(est_time=10, stage="base-b", runner_config="1-gpu-small")
```
Comment (Fridge003) [inline test/registered/unit/layers/test_layernorm_sp.py]: Please make it a cpu test

File: test/registered/models_e2e/test_layernorm_sp.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+register_cuda_ci(est_time=300, stage="base-b", runner_config="2-gpu-large")
+
+# Qwen3 dense (architecture "Qwen3ForCausalLM") is the SP allowlist entry.
+LAYERNORM_SP_MODEL = "Qwen/Qwen3-1.7B"
+
+
+class TestLayerNormSPAccuracy(CustomTestCase, GSM8KMixin):
+    gsm8k_score_threshold = 0.5
```
Comment (Fridge003) [inline test/registered/models_e2e/test_layernorm_sp.py]: Please use Qwen3-8B, and 0.85 as gsm8k score threshold

File: python/sglang/srt/layers/linear.py [hunk 7]
```diff
@@ -484,6 +485,18 @@ def weight_loader_v2(self, param: Parameter, loaded_weight: torch.Tensor):
     def forward(self, input_):
         bias = self.bias if not self.skip_bias_add else None
 
+        # Megatron SP "g": while the SP forward region is active, every
```
Comment (Fridge003) [inline python/sglang/srt/layers/linear.py]: Remove AI generated comments like this one. Or summarize them into 1-2 sentences

File: python/sglang/srt/runtime_context.py [hunk 8]
```diff
@@ -501,6 +501,11 @@ class ForwardFlags:
         "fuse_mlp_allreduce": False,
         "mlp_reduce_scatter": False,
         "flashinfer_trtllm_bypass": False,
+        # LayerNorm sequence parallelism (layers/layernorm_sp.py): set at the
```
Comment (Fridge003) [inline python/sglang/srt/runtime_context.py]: Remove ai comments

File: python/sglang/srt/server_args.py [hunk 9]
```diff
@@ -9847,6 +9857,18 @@ def check_server_args(self):
             self.dp_size > 1 and self.nnodes != 1 and not self.enable_dp_attention
         ), "multi-node data parallel is not supported unless dp attention!"
 
+        if self.enable_layernorm_sp:
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Please create a layernorm_sp_hook under args_group, and move the validation there. 
The validation should be a part of _run_resolution_pipeline rather then `check_server_args.py`

File: python/sglang/srt/layers/layernorm_sp.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    region (the exit gather) must use this rather than the ``sp_active`` flag:
+    Python writes inside the captured region do not re-execute on graph replay, so
+    the flag is stale there while this stays correct.
+    """
+    from sglang.srt.model_executor.forward_batch_info import ForwardMode
+
+    return layernorm_sp_enabled() and forward_mode == ForwardMode.EXTEND
+
+
+def should_activate_sp(forward_batch: ForwardBatch) -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/layers/layernorm_sp.py]: This function looks redundant

File: python/sglang/srt/layers/layernorm_sp.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+# The mechanism is generic; extend the allowlist as families are validated.
+_SP_SUPPORTED_ARCHITECTURES = frozenset({"Qwen3ForCausalLM"})
+
+
+def _model_architecture() -> Optional[str]:
+    architectures = get_server_args().get_model_config().hf_config.architectures
+    return architectures[0] if architectures else None
+
+
+def layernorm_sp_enabled() -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/layers/layernorm_sp.py:64]: It can instead be a global variable set only once at engine start

File: python/sglang/srt/layers/layernorm_sp.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    return layernorm_sp_enabled() and forward_mode == ForwardMode.EXTEND
+
+
+def should_activate_sp(forward_batch: ForwardBatch) -> bool:
+    """Whether this forward runs SP (see ``runs_sp``)."""
+    return runs_sp(forward_batch.forward_mode)
+
+
+# --- per-forward active flag (read at depth by linears / communicator) --------
+def is_sp_active() -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/layers/layernorm_sp.py]: This function is also redundant. We can use inline get_forward().sp_active instead

File: python/sglang/srt/layers/logits_processor.py [hunk 13]
```diff
@@ -356,6 +357,30 @@ def forward(
...
+        # model-agnostic; no-op when SP is inactive.
+        #
+        # The condition is recomputed from config + forward mode (runs_sp) instead
+        # of reading the sp_active flag, and the token count comes from input_ids:
+        # this code can sit outside the CUDA-graph-captured region, where the
+        # Python writes made inside it (prepare_attn's set_sp_active /
+        # set_sp_num_tokens) do not re-execute on replay and would read stale.
+        # Speculative aux capture is incompatible with SP (guarded off), so
+        # aux_hidden_states is never sharded.
+        if layernorm_sp.runs_sp(logits_metadata.forward_mode) and input_ids is not None:
```
Comment (Fridge003) [inline python/sglang/srt/layers/logits_processor.py]: Pack line 373-382 into a standalone function in layernorm_sp, and call it here

File: python/sglang/srt/layers/layernorm_sp.py (hunk 10 above)
Comment (MartinHua) [reply python/sglang/srt/layers/layernorm_sp.py]: deleted

File: python/sglang/srt/layers/layernorm_sp.py (hunk 12 above)
Comment (MartinHua) [reply python/sglang/srt/layers/layernorm_sp.py]: deleted

File: python/sglang/srt/runtime_context.py (hunk 8 above)
Comment (MartinHua) [reply python/sglang/srt/runtime_context.py]: trimmed

File: test/registered/models_e2e/test_layernorm_sp.py (hunk 6 above)
Comment (MartinHua) [reply test/registered/models_e2e/test_layernorm_sp.py]: done

File: test/registered/unit/layers/test_layernorm_sp.py (hunk 5 above)
Comment (MartinHua) [reply test/registered/unit/layers/test_layernorm_sp.py]: done

File: python/sglang/srt/layers/logits_processor.py (hunk 13 above)
Comment (MartinHua) [reply python/sglang/srt/layers/logits_processor.py]: New layernorm_sp.maybe_exit_gather(...)

File: python/sglang/srt/server_args.py (hunk 9 above)
Comment (MartinHua) [reply python/sglang/srt/server_args.py]: moved to layernorm_sp_hook

File: python/sglang/srt/layers/layernorm_sp.py (hunk 11 above)
Comment (MartinHua) [reply python/sglang/srt/layers/layernorm_sp.py:64]: created SpFlags

**Review by Fridge003 — APPROVED** [review]

**Review by Fridge003 — APPROVED** [review]
